File: JAVA_code/Lecture_23.java
                                                //LECTURE-23
                                                //july-

import java.util.*;
public class Lecture_23
{   
    //Pre-requist for dealing with graph.
    public static class Edge {
        int nebr = -1;
        int wt = -1;

        Edge(int n, int w) {
            this.nebr = n;
            this.wt = w;
        }
    }
    static ArrayList<ArrayList<Edge> > graph = new ArrayList<>();

    
    static void addEdge(int u, int v, int wt) {
        graph.get(u).add(new Edge(v, wt));
        graph.get(v).add(new Edge(u, wt));
    }


    static void display() {
        for(int i = 0; i < graph.size(); i++) {
            System.out.print(i + " -> ");
            for(int j = 0; j < graph.get(i).size(); j++) {
                System.out.print( "[ " + graph.get(i).get(j).nebr + ", " + graph.get(i).get(j).wt + "], ");
            }
            System.out.println();
        }
    }

    //REVIEWED.
    
    

    //1. BFS traversal
    
    static boolean bfs(int src, int des, boolean[] visited) {      //here, instead of making visited global,
        Queue<Integer> que = new LinkedList<>();                     //we have passed visited array in function
        que.add(src);                                                //arguments, so during it's size declaration
                                                                     //i dont need to define its size as constant
        while(que.isEmpty() == false) {                              //, as we have to do in cpp.
            int rem = que.poll();           //get and remove

            visited[rem] = true;            //mark

            if(rem == des) {                //work
                return true;
            }

            for(int i = 0; i < graph.get(rem).size(); i++) {     //add nebr.
                Edge e = graph.get(rem).get(i);
                if(visited[e.nebr] == false) {
                    que.add(e.nebr);
                }
            }
        }
        
        return true;
    }

    //REVIEWED.



    

    //2. GET_CONNECTED_COMPONENTS for a graph

    static String gscc(int s, boolean[] visited) {
        String str = "";
        Queue<Integer> que = new LinkedList<>();
        que.add(s);
        
        while(que.isEmpty() == false) {
            int rem = que.poll();           //GET and REMOVE at the same time

                                            //don't need to mark, as already done proactively,(so 
                                            //no duplicate gets added to queue)

            str += rem;                    //WORK

            for(int i = 0; i < graph.get(rem).size(); i++) {      //ADD NEBR
                Edge e = graph.get(rem).get(i);
                if(visited[e.nebr] == false) {
                    visited[e.nebr] = true;
                    que.add(e.nebr);
                }
            }
        }

        return str;
    }



    static void gcc(boolean[] visited) {
        ArrayList<String> components = new ArrayList<>();

        for(int i = 0; i < graph.size(); i++) {
            if(visited[i] == false) {
                visited[i] = true;
                String str = gscc(i, visited);
                components.add(str);
            }
        }

        for(String x : components) {
            System.out.print(x + ", ");
        }

    }
    
    //REVIEWED.
    







    //3. ROTTEN_ORANGES_PROBLEM.               (implementation of bfs)

    static class point {
        int x = -1;
        int y = -1;
        int t = -1;

        point(int x, int y, int t) {
            this.x = x;
            this.y = y;
            this.t = t;
        }
    }


    static void rottenOranges(int[][] board, int time) {
        Queue<point> que = new LinkedList<>();
        
        for(int i = 0; i < board.length; i++) {
            for(int j = 0; j < board[0].length; j++) {
                if(board[i][j] == 2) {
                    point p = new point(i, j, 0);
                    que.add(p);   
                }
            }
        }

        while(que.isEmpty() == false) {
            point p1 = que.poll();              //get and remove

            //here, work is done prior to marking. 
            if(p1.t == time + 1){               //WORK
                break;
            }

            board[p1.x][p1.y] = 2;              //mark

            if(p1.x - 1 >= 0 && board[p1.x - 1][p1.y] == 1) {          //add nebr
                que.add(new point(p1.x - 1, p1.y, p1.t + 1));
            }
            if(p1.y + 1 < board[0].length && board[p1.x][p1.y + 1] == 1) {
                que.add(new point(p1.x, p1.y + 1, p1.t + 1));
            }
            if(p1.x + 1 < board.length && board[p1.x + 1][p1.y] == 1) {
                que.add(new point(p1.x + 1, p1.y, p1.t + 1));
            }
            if(p1.y - 1 >= 0 && board[p1.x][p1.y - 1] == 1) {
                que.add(new point(p1.x, p1.y - 1, p1.t + 1));
            }
            
        }

        for(int i = 0; i < board.length; i++) {               //printing conditon of board after time.
            for(int j = 0; j < board[0].length; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }

    }

    //REVIEWED.
    
    






    public static void main(String[] args)
    {   
        //pre-requist for graph
        for(int i = 0; i < 7; i++) {
            graph.add(new ArrayList<>());
        }
        
        addEdge(0,1,10);
        addEdge(1,2,10);
        addEdge(2,3,10);
        addEdge(0,3,40);
        addEdge(2,5,5);            //extra edge between 2,5.
        addEdge(3,4,2);
        addEdge(4,5,3);
        addEdge(5,6,3);
        addEdge(4,6,8);
    
        boolean[] visited = new boolean[graph.size()];
        //reviewed.



        //1. BFS searching.
        
        /*int src = 0;
        int des = 6;
        System.out.println(bfs(src, des, visited));
        */

        //reviewed.





        //2. GET_CONNECTED_COMPONENTS                   (implementation of bfs)

        /*
        gcc(visited);
        */

        //reviewed.





        //3. ROTTEN_ORANGES_PROBLEM                    (implementation of bfs)
        
        /*
        int[][] board = {
            {2,1,0,2,1},
            {1,0,1,2,1},
            {1,0,0,2,1}
        };

        int time = 1;
        rottenOranges(board, time);
        */

        //reviewed.

    }


}

File: C++_code/Lecture_23.cpp
                                            //LECTURE-23
                                            //july-19

#include<iostream>
#include<vector>
#include<string>
#include<queue>
using namespace std;


//PRE- requist for graph

class Edge                                   
{                                               
    public:                                     
        int nebr;                               
        int wt; 
};

vector<vector<Edge>> graph;             //DATA TYPE TO REPRESENT GRAPH.

void addEdge(int v1,int v2, int wt)                    //it will add the edge in the graph data
{                                                    
    Edge e1;
    e1.nebr=v2;
    e1.wt=wt;
    graph[v1].push_back(e1);

    Edge e2;
    e2.nebr=v1;
    e2.wt=wt;
    graph[v2].push_back(e2);    
}

void display() {
    for(int i = 0; i < graph.size(); i++) {
        cout<<i<<" -> ";
        for(int j = 0; j < graph[i].size(); j++) {
            cout<<"[ "<<graph[i][j].nebr<<", "<<graph[i][j].wt<<"] ";
        }
        cout<<"\n";
    }
}

//REVIEWED.





//1. BFS traversal, over the graph.

vector<bool> visited (7, false);               //we have to give const. size to visited vector, as
bool BSF_traversal(int src,int des)            //if we use graph.size(), it may not have initialize it's size.
{
    queue<int> q;
    q.push(src);
    while(q.size()>0)
    {
        int rem=q.front();          //GET
        
        q.pop();                    //REMOVE
        
        visited[rem]=true;          //MARK

        if(rem==des)                //WORK
            return true;

        for(int i=0;i<graph[rem].size();i++)     //ADD NEIGHBOUR.
        {
            Edge n=graph[rem][i];                  //here, some of vertices will be added multiple times.
            if(visited[n.nebr]==false)
                q.push(n.nebr);
        }
    }

    return false;
}

//REVIEWED.







//2. Get-Connected-Components.     (for a graph), (gscc is a helping function)

string gscc(int i,vector<bool> &visited)        //Get-selected-connnected-conponents.
{   
    string res="";
    queue<int> q;
    visited[i] = true;
    q.push(i);

    while(q.size()>0)
    {
        int rem=q.front();       //get
        q.pop();                 //remove
                                 //No need to mark, as we have done marking pro-actively.

        res+=to_string(rem);     //work

        for(int j=0;j<graph[rem].size();j++)      //add nebr
        {
            Edge n=graph[rem][j];
            if(visited[n.nebr]==false) {
                visited[n.nebr] = true;
                q.push(n.nebr);
            }
        }
    }
    return res;
}



vector<string> gcc()            //actual function to get connected components.
{
    vector<string> comps;
    vector<bool> visited(graph.size(),false);

    for(int i=0;i<graph.size();i++)
    {
        if(visited[i]==false)
        {
            string comp=gscc(i,visited);
            comps.push_back(comp);
        }
    }
    return comps;
}

//REVIEWED.







//3. FIRE_STROM_PROBLEM        (implementation of bfs) 

class point {
    public:
        int x = -1;
        int y = -1;
        int t = -1;

        point(int x, int y, int t) {
            this -> x = x;
            this -> y = y;
            this -> t = t;
        }
};


void fireStrom(int board[][5], int rSize, int cSize) {
    queue<point> que;                                   //here, que will have multiple dupliacte
    for(int i = 0; i < rSize; i++) {                    //items, as we did'nt mark and check pro-actively.
        for(int j = 0; j < cSize; j++) {
            if(board[i][j] == 2) {
                point p1(i, j, 0);
                que.push(p1);
            }
        }
    }    

    int totalTime = -1;          //variable to get total time for buring max.

    while(que.size() != 0) {
        point p = que.front();                 //GET
        que.pop();                             //REMOVE

        board[p.x][p.y] = 2;                    //MARK

        totalTime = p.t;                       //work 

        if(p.x - 1 >= 0 && board[p.x - 1][p.y] == 1) {              //add nebrs
            point temp(p.x - 1, p.y, p.t + 1);
            que.push(temp);
        }
        if(p.y + 1 < cSize && board[p.x ][p.y + 1] == 1) {
            point temp(p.x, p.y + 1, p.t + 1);
            que.push(temp);
        }
        if(p.x + 1 < rSize && board[p.x + 1][p.y] == 1) {
            point temp(p.x + 1, p.y, p.t + 1);
            que.push(temp);
        }
        if(p.y - 1 >= 0 && board[p.x][p.y - 1] == 1) {
            point temp(p.x, p.y - 1, p.t + 1);
            que.push(temp);
        }

    }

    cout<<"Time requried for max burning = "<<totalTime<<"\n";

    for(int i = 0; i < rSize; i++) {             //printing the last condition of board.
        for(int j = 0; j < cSize; j++) {
            cout<<board[i][j]<<" ";
        }
        cout<<"\n";
    }
}

//REVIEWED.






int main()
{
    //pre-requist for graph

    graph.push_back(vector<Edge>()); //0
    graph.push_back(vector<Edge>()); //1
    graph.push_back(vector<Edge>()); //2
    graph.push_back(vector<Edge>()); //3
    graph.push_back(vector<Edge>()); //4
    graph.push_back(vector<Edge>()); //5
    graph.push_back(vector<Edge>()); //6

    addEdge(0,1,10);
    addEdge(1,2,10);
    addEdge(2,3,10);
    addEdge(0,3,40);
    addEdge(3,4,2);
    // addEdge(2,5,5);
    addEdge(4,5,3);
    addEdge(5,6,3);
    addEdge(4,6,8);

    // display();
    
    //reviewed.



    //1. BFS searching.
    
    /*
    cout<<BSF_traversal(0,6)<<"\n";
    */

   //reviewed.
    



    //2. GET_CONNECTED_COMPONENTS of a graph. 
    
    /*vector<string> comps = gcc();
    for(int i=0;i<comps.size();i++)
    {
        cout<<comps[i]<<endl;
    }
    */

    //reviewed.



    
    //3. FIRE_STROM_PROBLEM               (implementation of bfs) 
    
    /*
    int board[][5] = {
        {2,1,0,2,1},
        {1,0,1,2,1},
        {1,0,0,2,1}
    };
    fireStrom(board, 3, 5);
    */

    //reviewed.




}
